task/net/src/vlan.rs

```diff
@@ -0,0 +1,407 @@
+                VLanTxToken(self.eth, self.vid),
+            ))
+        } else {
+            None
+        }
+    }
+    fn transmit(&'a mut self) -> Option<Self::TxToken> {
+        if self.eth.can_send() {
+            Some(VLanTxToken(self.eth, self.vid))
+        } else {
+            None
+        }
+    }
+    fn capabilities(&self) -> smoltcp::phy::DeviceCapabilities {
+        self.eth.capabilities()
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+pub struct VLanRxToken<'a>(&'a eth::Ethernet, u16);
+impl<'a> smoltcp::phy::RxToken for VLanRxToken<'a> {
+    fn consume<R, F>(
+        self,
+        _timestamp: smoltcp::time::Instant,
+        f: F,
+    ) -> smoltcp::Result<R>
+    where
+        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
+    {
+        self.0.vlan_recv(self.1, f)
+    }
+}
+
+pub struct VLanTxToken<'a>(&'a eth::Ethernet, u16);
+impl<'a> smoltcp::phy::TxToken for VLanTxToken<'a> {
+    fn consume<R, F>(
+        self,
+        _timestamp: smoltcp::time::Instant,
+        len: usize,
+        f: F,
+    ) -> smoltcp::Result<R>
+    where
+        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
+    {
+        self.0
+            .vlan_try_send(len, self.1, f)
+            .expect("TX token existed without descriptor available")
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// State for the running network server
+pub struct ServerImpl<'a> {
+    eth: &'a eth::Ethernet,
+
+    socket_handles: [[SocketHandle; SOCKET_COUNT]; VLAN_COUNT],
+    ifaces: [Interface<'a, VLanEthernet<'a>>; VLAN_COUNT],
+    bsp: crate::bsp::Bsp,
+}
+
+impl<'a> ServerImpl<'a> {
+    /// Size of buffer that must be allocated to use `dispatch`.
+    pub const INCOMING_SIZE: usize = idl::INCOMING_SIZE;
+
+    /// Builds a new `ServerImpl`, using the provided storage space.
+    pub fn new(
+        storage: &'a mut ServerStorage<'a>,
+        mut ipv6_addr: Ipv6Address,
+        mut mac: EthernetAddress,
+        bsp: crate::bsp::Bsp,
+    ) -> Self {
+        // Local storage; this will end up owned by the returned ServerImpl.
+        let mut socket_handles = [[Default::default(); generated::SOCKET_COUNT];
+            generated::VLAN_COUNT];
+        let mut ifaces: [Option<Interface<'a, VLanEthernet<'a>>>; VLAN_COUNT] =
+            Default::default();
+
+        // We're iterating over a bunch of things together.  The standard
+        // library doesn't have a great multi-element zip, so we'll just
+        // manually use mutable iterators.
+        let mut neighbor_cache_iter = storage.neighbor_cache_storage.iter_mut();
+        let mut socket_storage_iter = storage.socket_storage.iter_mut();
+        let mut socket_handles_iter = socket_handles.iter_mut();
+        let mut vid_iter = generated::VLAN_RANGE;
+        let mut ifaces_iter = ifaces.iter_mut();
+        let mut ip_addr_slice = &mut storage.ipv6_net[..];
```

Comment:
I think this could be

```rust
let mut ip_addr_iter = storage.ipv6_net.chunks_mut(1);
```

to allow it to be an iterator like the others instead of having to manually carve it up via `.split_at_mut(1)`.

Reply:
Nice!